src/proto/streams/prioritize.rs

```diff
@@ -476,122 +484,142 @@ where
-
-                            // Decrement the stream's buffered data counter
-                            debug_assert!(stream.buffered_send_data >= len);
-                            stream.buffered_send_data -= len;
-                            stream.requested_send_capacity -= len;
-
-                            // Assign the capacity back to the connection that
-                            // was just consumed from the stream in the previous
-                            // line.
-                            self.flow.assign_capacity(len);
-
-                            trace!(" -- updating connection flow --");
-                            self.flow.send_data(len);
-
-                            // Wrap the frame's data payload to ensure that the
-                            // correct amount of data gets written.
-
-                            let eos = frame.is_end_stream();
-                            let len = len as usize;
+                counts.transition_after(stream, is_counted);
 
-                            if frame.payload().remaining() > len {
-                                frame.set_end_stream(false);
-                            }
+                return Some(frame);
+            }
 
-                            Frame::Data(frame.map(|buf| {
-                                Prioritized {
-                                    inner: buf.take(len),
-                                    end_of_stream: eos,
-                                    stream: stream.key(),
-                                }
-                            }))
-                        },
-                        frame => frame.map(|_| unreachable!()),
-                    };
-
-                    trace!("pop_frame; frame={:?}", frame);
-
-                    if !stream.pending_send.is_empty() {
-                        // TODO: Only requeue the sender IF it is ready to send
-                        // the next frame. i.e. don't requeue it if the next
-                        // frame is a data frame and the stream does not have
-                        // any more capacity.
-                        self.pending_send.push(&mut stream);
+            // check for any pending open streams
+            if counts.can_inc_num_send_streams() {
+                while let Some(mut stream) = self.pending_open.pop(store) {
+                    trace!("pop_frame; pending_open {:?}", stream.id);
+                    if let Some(frame) = stream.open_headers.take() {
+                        let end_stream = frame.is_end_stream();
+                        let mut task = stream.open_task.take();
+                        // Update the state
+                        if let Err(_e) = stream.state.send_open(end_stream) {
+                            // other peer may have closed this stream between us pending
+                            // and finally opening
+                            error!("pending open errored");
```

Comment:
This may not be great. I originally had `send_open(..).expect("this shouldn't happen")`...